File: 20. Valid Parentheses/main.py
class Solution:
    def isValid(self, s: str) -> bool:
        stck = []
        opening = ['(', '[', '{']
        closing = [')', ']', '}']
        for ch in s:
            if ch in opening:
                stck.append(ch)
            elif ch in closing:
                if not len(stck):
                    return False
                last = stck.pop()
                index = closing.index(ch)
                if last != opening[index]:
                    return False
        
        return len(stck) == 0
                
            
        

File: 20. Valid Parentheses/main.js
/**
 * @param {string} s
 * @return {boolean}
 */
 var isValid = function (s) {
    let stck = []
    let valid = true;
    
    for (let i = 0; i < s.length && valid; i++) {
        let ch = s[i];

        if (ch === '(' || ch === '{' || ch === "[") {
            stck.push(ch);
        } else if (ch === ')' || ch === ']' || ch === '}') {

            if (!stck.length) {
                valid = false;
            }

            let last = stck.pop();
            switch (ch) {
                case ')': {

                    if (last != '(') {
                        valid = false
                    }
                    break;
                }
                case ']': {
                    if (last != '[') {
                        valid = false;
                    }
                    break;
                }
                case '}': {
                    if (last != '{') {
                        valid = false;
                    }
                    break;
                }

            }
        }
    }

    if (stck.length)
        return false;
    return valid;
};